docs/content/sips/011-component-versioning.md

```diff
@@ -0,0 +1,56 @@
+template = "main"
+date = "2023-01-24T01:01:01Z"
+---
+
+Summary: Embed version information into Spin components for improved compatibility diagnostics.
+
+Owner(s): joel.dice@fermyon.com
+
+Created: Jan 24, 2023
+
+## Background
+
+Currently, Spin has no way to determine which version of SDK the components of an application were built with, which leads to obscure, hard-to-understand error messages for ABI and/or API incompatibilities.  For example, a component built with a Spin 0.4 SDK won't run on Spin 0.7 due to `wit-bindgen` ABI changes, leading to a variety of confusing runtime errors like out-of-bounds memory traps, assertion errors about invalid values, etc.  In addition, new APIs were added between 0.4 and 0.7, and thus a 0.7 component might not run on 0.4 even if the ABI had not changed.  These errors are usually a bit easier to understand (e.g. missing imports), but still don't point clearly to a resolution.  Ideally, the message presented to the user would be something like, "This component targets Spin 0.7 and cannot be run using this version of Spin (0.4) -- please use Spin 0.7.x to run it."
+
+## Proposal
+
+Although `WIT` does not yet have a way to express API versions, we can emulate it by exporting a set of core WASM functions from the SDK (and thus any component built using that SDK) using the following naming convention:
+
+```
+spin-sdk-version-$MAJOR-$MINOR
+spin-sdk-language-$LANGUAGE
+spin-sdk-commit-$HASH
+```
+
+where `$MAJOR` and `$MINOR` are the Spin major and minor version numbers the SDK targets, `$LANGUAGE` is the programming language the SDK supports, and `$HASH` is the Git commit hash the SDK was built from. The intention behind embedding this information in the name of the function is that it can be checked statically without instantiating or running the component, and it is immune to component model ABI changes.
+
+For languages which don't yet have an version-enabled Spin SDK (or developers who wish to use `wit-bindgen` directly, or even implement bindings by hand), the component may itself export any or all of the above functions.
+
+### Targetting unreleased Spin versions
+
+The above works fine for Spin releases, but we might also want to build components that target pre-release builds of Spin, which may include an ABI and/or API that's different from what ends up in the next release, and yet also different from what was in the previous release.  To handle this case, we propose adding a step to the Spin release process such that the version stored in the `workspace.package.version` field of Spin's Cargo.toml file is changed to `$MAJOR.$MINOR.0-pre0`, where `$MAJOR.$MINOR.0` is the anticipated next Spin release.  Then, each time an ABI or API change is made, the `-pre$N` suffix is incremented.  For example, when Spin 0.9 is released, the version on the main branch is updated to `0.10.0-pre0`, and if an API change is made a few days later, it's updated again to `0.10.0-pre1`.
```

Comment:
Curious as to why the patch version is hard-coded to `0`.  Does this mean if spin or and SDK is on a non-zero patch release, eg 0.9.1, does the `workspace.package.version` need to remain `0.9.0-pre$N`?  Or should this be reworded to allow for arbitrary patch values, eg `$MAJOR.$MINOR.$N-pre$N`?

Reply:
Good question.  We've had some discussion with Luke Wagner and others about what role patch version numbers should play in Component Model interface versioning (which this SIP is designed to anticipate), and the current thinking is that it's not relevant.  Versioning interfaces is different than versioning implementations, and while patch versions definitely make sense for the latter, they're not necessarily useful for the former.

Also, the _anticipated_ next Spin release should always have a zero patch version, since non-zero patch versions are for _unanticipated_ issues.